### ISSUE TITLE
Metrics for object detection with bounding boxes

### ISSUE THREAD

**Issue by tadejsv** [PyTorchLightning/pytorch-lightning#4988]
## 🚀 Feature

We could add a whole bunch of metrics for object detection (bounding boxes) by just writing a plugin that transforms the bounding boxes to multi-class label tensor representation.

### Motivation & Pitch

Metrics for bounding box detection are basically just classification metrics (see [here](https://github.com/rafaelpadilla/Object-Detection-Metrics) for example) - the only difference is that the inputs come as bounding boxes which then have to be interpreted as class predictions.

If we write a plugin for this "interpretation" part - and I don't even think this would be that hard - we can just plug it into the classification metrics that we already have, and voila, we suddenly cover all object detection metrics!

I think it would still be good idea to create separate classes for these metrics (`ObjectDetectionAveragePrecision` or something?), but on the inside they would be a simple call to the transformation plugin + usual classification metrics.

If people think this is a good idea, I am willing to create this transformation plug-in and then slowly add these metrics :)

**Comment by SkafteNicki** [comment]
Maybe we can get the opinion on this from @briankosw, @LucaButera, @AlexTS1980 as they are all involved in this PR https://github.com/PyTorchLightning/pytorch-lightning/pull/4564 that tries to implement some metrics for object detection

**Comment by LucaButera** [comment]
@tadejsv I am not totally sure on this. Can you elaborate more on what you mean when you say that they are the same metrics but the only thing changing is the input format?
If you could provide an example that would be great.

**Comment by tadejsv** [comment]
@LucaButera yes, absolutely.

# Example

So consider this simple example (for one image):
``` python
# preds have the format [class, probability, x_min, y_min, x_max, y_max]
preds = [
    [0, 0.2, 10, 10, 110, 110],
    [0, 0.7, 50, 50, 150, 150],
    [1, 0.2, 300, 300, 350, 350],
    [1, 0.6, 50, 200, 150, 300],
    [1, 0.8, 0, 250, 100, 350],
]

# target have the format [class, x_min, y_min, x_max, y_max]
target = [
    [0, 30, 30, 130, 130],
    [1, 25, 225, 125, 325],
]
```
Obviously, IRL predictions would contain probability for each class (and some "not relevant" class) - here I'm simplifying things for this example.

Then, the hardest step (and the one which would be handled by the transform function) is how to match preds and target. I'm going to use the criteria from [here](https://github.com/rafaelpadilla/Object-Detection-Metrics#an-ilustrated-example), where they use the IoU 0.3 threshold for TP, and if there are 2 or more preds over the IoU threshold for one target, the one with the highest probability (confidence) is taken. 

Obviously, we could use different criteria, or even use some filtering like non-max suppression first. Here's how the inputs would be transformed:

``` python
>>> preds_tr, target_tr = object_detection_input_transform(preds, target)
>>> preds_tr
[
    [0.2, 0],
    [0.7, 0],
    [0, 0.2],
    [0, 0.6],
    [0, 0.8]
]
>>> target_tr
[
    [0, 0], #target was 0, IoU was 0.47, prob 0.2 
    [1, 0], #target was 0, IoU was 0.47, prob 0.7 
    [0, 0], #target was 1, IoU was 0, prob 0.2
    [0, 0], #target was 1, IoU was 0.39, prob 0.6
    [0, 1], #target was 1, IoU was 0.39, prob 0.8
]
```

Now our object detection problem was transformed into a multi-label classification problem. And from here on we can insert this into our standard classification metrics! In fact let's use sklearn`s `average_precision_score` with the `macro` setting to calculate mAP for this example (correct me if I am wrong and this actually calculates something else):
``` python
>>> average_precision_score(target_tr, preds_tr, average='macro')
1.0
```
# Conclusion

So I used sklearn in this example but we would use our own PL metrics. So I believe that the proper way to add object detection metrics is:
1. Create a function that tranforms inputs from object-detection format to standard classification problem format
2. Create a metric for classification inputs (like average precision score)
3. Combine the two (formatting function + classification metric) to create an object detection metric

Otherwise we would be repeating the metric calculation code all over the place - instead of using already well written and tested classification metrics.

**Comment by LucaButera** [comment]
@tadejsv It seems good reasoning. I am not 100% convinced and I would run some comparison tests to ensure that, with this approach, output values are correct in both common and corner cases. 

For instance, considering your example, I would not expect mAP to be ```1.0``` as each "extra" prediction should count as a False Positive.

This can be easily tackled by directly assigning a binary prediction score to the chosen class for the box. I think the catch here is that even if the confidence is ```0.2``` it still means that was the predicted class, while in you example it gets considered as a ```0```prediction. But I may be wrong on this. I don't know the internals of the function you used.

If I am correct your example should be transformed into something like:
``` python
>>> preds_tr, target_tr = object_detection_input_transform(preds, target)
>>> preds_tr
[
    [1, 0],
    [1, 0],
    [0, 1],
    [0, 1],
    [0, 1]
]
>>> target_tr
[
    [0, 0], #target was 0, IoU was 0.47, prob 0.2 
    [1, 0], #target was 0, IoU was 0.47, prob 0.7 
    [0, 0], #target was 1, IoU was 0, prob 0.2
    [0, 0], #target was 1, IoU was 0.39, prob 0.6
    [0, 1], #target was 1, IoU was 0.39, prob 0.8
]
```

Does this seems reasonable to you? 
In the end, if this approach produces consistent values if compared to known implementations, I think it could be a viable solution. I don't know if it is computationally as efficient as having custom metrics though.

**Comment by tadejsv** [comment]
@LucaButera Regarding mAP value, I think it is equal to 1.0 as all FP in my artificial inputs have confidence (probability) below or equal to that of the TP input. Please let me know if you can think of a more realistic example, and we can compare "my way" of doing this to the implementation in your PR

Regarding the transform you proposed - it's not correct. The reason that `preds_tr` gets assigned a probability and are not binarized is precisely so that mAP function would work - as you need to evaluate things at different thresholds. Oh and "the function" I used was so far just my reasoning/hand calculations :). I haven't started developing anything yet - I first opened this issue to gather some feedback.

And regarding computational efficiency: it is just as efficient as the standard implementation. Why? The computationally heavy part is assigning targets to predictions - and this part (that would be taken care of by `object_detection_input_transform`) is equal to the standard object-detection implementation. Then, all that's left is to construct the preds and target tensors (this is a single loop over all predictions), and calculation of the metrics themselves are again just as computationally efficient as object detection case, as the same algorithms are used. This is precisely the reason I am proposing this split into formatting and metrics :)

**Comment by LucaButera** [comment]
@tadejsv Yes I think you're right, as long as you get a matching box for each target your mAP will be ```1.0``` since the one considered TP is always the one with highest confidence and there's only one image in this case. 
You could compare with [these](https://github.com/PyTorchLightning/pytorch-lightning/blob/92483a8a0302eee0e48964898e89c217fcee20ae/tests/metrics/functional/test_classification.py#L357-L401) test cases, to see if the results are correct.

How do you plan to handle multiple images? Do you just add up to the targets so that each row represent the "artificial" target for a particular box in a particular image? I thought about that and intuitively it should work as well.

The only thing that bugs me is that this mapping from detection outputs to multi class classification outputs is not guaranteed to be the correct one for each possible metric, given that the metric definition is arbitrary (for example if you where to consider COCO's box size constrained mAP, the mapping would be different even though similar).

In the end I think that, given a reasonable number of tests, possibly on both random inputs and hand tailored corner cases (no matching box, all boxes perfectly matching...) this sure causes no harm. 
A detection metric would consist of a ```mapping``` and a ```metric``` part and, as long as this is transparent to the model, it may split the concerns in a graceful manner.

On the other hand, if we end up having as many "mappings" as metrics we wish to implement, then I think this may make the code less readable. Could we think of other commonly used detection metrics and try and understand how much this solution cleans our "future" code?

**Comment by tadejsv** [comment]
Thanks, I'll try it out on the test cases later today.

About multiple images: what you propose should work yes - this way of aggregation (aggregating boxes across all images) is also what I have seen done elsewhere. Additionally, I could introduce an option to compute the metric for each image, and average across images. This would tie into some other PRs I have open for classification metrics at the moment. Although I am not sure if this method of aggregation is used for object detection at all.

So regarding different mappings: I think that this would not make the code less readable. We could simply create a separate function for each mapping. I mean, I don't think there is any alternative that can somehow bypass writing different mappings; or one that makes it all possible with less code.

I agree that we need to riguorously test all different edge cases and similar. And about other metrics: from what I have seen, all that is used is some version of mAP, or the area under the precision-recall curve. If you can think of anything else, let me know

**Comment by LucaButera** [comment]
@tadejsv Could you point me to some examples of where you saw this approach implemented? I am really interested since I have only been able to find "raw" approaches. 
I am not sure if computing the metric for each image is interesting for detection, but in the end it is just an additional feature so, as long as we keep it tidy, it's a bonus.

> And about other metrics: from what I have seen, all that is used is some version of mAP, or the area under the precision-recall curve. If you can think of anything else, let me know

See that is something that got me thinking. Since detection uses mostly these metrics (I can't think of other widely used ones aside from Average Recall which I am not really familiar with) the advantage of having this plugin instead of a custom metric is really marginal, so we need to have reasonable proof that this approach is "mathematically" the same, cause if you screw the metrics, well, you screw everything. Of course I am not speaking of a rigorous proof, but a wide enough tests' set.

Another thing I came up with: do the ```metrics``` components that we're gonna use implement the computation of AP with interpolation? Commonly used mAP metrics (COCO, PascalVOC) compute the P/R curve over interpolated values,(I think in order to reduce the impact of class population differences). As far as I can see PL Metrics do not support this. We have to take this into account as we should either modify the existing metrics or write new ones (which I think defeats the purpose).

In the end one thing I really like of this approach is that decoupling the mapping from the metric computation should allow to implement many of the mAP variants easily, as the plugin may do all the work to split boxes by aspect ratio, size and whatnot. Take a look at [COCO Evaluation Specification](https://cocodataset.org/#detection-eval) if you haven't already.

**Comment by tadejsv** [comment]
@LucaButera  Apologies for the late reply. So I tried my method on the first example from the test - and now I am actually pretty sure that "expected" score of the example is just wrong. Let me demonstrate:

``` python
import torch
from torchvision.ops import box_iou
from sklearn.metrics import average_precision_score

preds = torch.tensor([
    [0, 0.6, 62, 83, 225, 195],
    [0, 0.1, 79, 93, 118, 131],
    [0, 0.9, 117, 192, 127, 244],
    [0, 0.2, 15, 201, 26, 254],
    [0, 0.65, 35, 45, 210, 170],
    [0, 0.7, 210, 80, 295, 90]
])
target = torch.tensor([
    [0, 0, 100., 50., 205., 200.],
    [0, 0, 85., 83., 225., 195.],
    [0, 0, 117., 192., 127., 244.],
    [0, 0, 10., 200., 15., 254.],
    [0, 0, 30., 40., 234., 150.],
    [0, 0, 210., 80., 295., 90.],
])
```

So then, let's match predictions to targets - we match each prediction to the target with which it has the highest IoU. This is done like so:
```python
>>> ious = box_iou(preds[:, 2:], target[:, 2:])
>>> torch.max(ious, dim = 1)
... torch.return_types.max(
... values=tensor([0.8589, 0.0788, 1.0000, 0.0000, 0.7084, 1.0000]),
... indices=tensor([1, 1, 2, 0, 4, 5]))
```

We see that the unmatched predictions (with IoU below 0.5) are exactly those ones with the lowest confidences (`0.1` and `0.2`).  I think from this it is clear that mAP should be acutally 1 - and not 2/3 as the test claims it should be. Precision always equals 1 here.

But to drive the point home, let's complete the example:
``` python
>>> target_tr = (torch.max(ious, dim = 1).values >= 0.5).reshape(6,1)
>>> preds_tr = preds[:, [1]]
>>> average_precision_score(target_tr.numpy(), preds_tr.numpy(), average='macro')
1.0
```

**Comment by tadejsv** [comment]
> Could you point me to some examples of where you saw this approach implemented? I am really interested since I have only been able to find "raw" approaches.

Sorry, when I said that I have seen that way of aggregation elsewhere I didn't mean I saw it implemented my way, just that it was done that way, as opposed to calculating the score for each image and averaging. I haven't seen the metric implemented in mapping/metric way, unfortunately.

> See that is something that got me thinking. Since detection uses mostly these metrics (I can't think of other widely used ones aside from Average Recall which I am not really familiar with) the advantage of having this plugin instead of a custom metric is really marginal, so we need to have reasonable proof that this approach is "mathematically" the same, cause if you screw the metrics, well, you screw everything. Of course I am not speaking of a rigorous proof, but a wide enough tests' set.

I mean, to me it is very clear that the approach is the same. Would love to prove, but would need some correct examples first haha. But as you say later yes, perhaps the real advantage of this approach is not that it can be used for many metrics (as apparently only 1/2 are used), but that it can be used for many different mappings. This seems as important, if not more, than the first option.

> Another thing I came up with: do the metrics components that we're gonna use implement the computation of AP with interpolation? Commonly used mAP metrics (COCO, PascalVOC) compute the P/R curve over interpolated values,(I think in order to reduce the impact of class population differences). As far as I can see PL Metrics do not support this. We have to take this into account as we should either modify the existing metrics or write new ones (which I think defeats the purpose).

I think probably the best approach would be then to add the interpolation option to PL metrics first, and then to write object detection metrics.

> In the end one thing I really like of this approach is that decoupling the mapping from the metric computation should allow to implement many of the mAP variants easily, as the plugin may do all the work to split boxes by aspect ratio, size and whatnot. Take a look at COCO Evaluation Specification if you haven't already.

Thank you for the link, I'll take a look at it.

**Comment by LucaButera** [comment]
@tadejsv I agree with what you said, except for the result correctness. If I run that example against my implementation (which is not the one in the PR where the example comes from) I get ```0.6667``` which is indeed ```~2/3```. I tested my implementation against ```pycocotools``` and it should be correct. So either I did something wrong somewhere or there's a catch in your solution that we can't quite grasp.

The only way to have a final answer (or at least the only I see that doesn't require a rigorous proof) is to write a script that randomly generates bounding boxes and translates them both into [pycocotools](https://github.com/cocodataset/cocoapi/tree/master/PythonAPI/pycocotools)'s representation and yours. Then it compares the results computed by the two methods.
Right now I don't have time to do so, but I guess it is the only way to assess the reliability of the idea, as ```pycocotools``` is, I hope you agree, a trustworthy baseline.

**Comment by tadejsv** [comment]
@LucaButera Actually, you're right. I figured out why my version doesn't work: there are two target objects that are not matched by any of the preds objects. This creates 2 extra FN entries that influence Recall. To model this in my way would require adding two fictitious predictions with 0 confidence and then cutting off the precision-recall curve at the lowest actual confidence (similar to the `max_fpr` for ROC curve). 

But yes, seems that just using base classification metrics would not be possible, and some modifications would be needed. Although I still think that mapping/metric approach would work well. Anyways, as it seems my original idea is not directly implementable, I'm closing this issue

**Comment by sick78** [comment]
i’m a teacher we all make mistakes but we learn n growin is the point n i have grown when i am in an webinar called to order put me on ther camera cuts n puts me live ok n send the traffic cool

**Comment by yipliu** [comment]
So what should PL do about the Bbox case?
